### PR TITLE
initial version of a config fetcher.

### PR DESCRIPTION
**Pull request by samohte**
initial version of a config fetcher.
At the moment, it queries the span endpoint every minute. 
I had to extend the sampler factory interface to allow to re-use the tracer's HttpService in the sampler.

### CODE REVIEW

File: source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_fetcher.cc [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    } else {
+      Http::RequestMessagePtr message = Http::Utility::prepareHeaders(http_service_.http_uri());
+      // TODO: set path
+      // message->headers().setPath("path/to/sampler/service");
+      message->headers().setReferenceMethod(Http::Headers::get().MethodValues.Post);
+      // Add all custom headers to the request.
+      for (const auto& header_pair : parsed_headers_to_add_) {
+        message->headers().setReference(header_pair.first, header_pair.second);
+      }
+      active_request_ = thread_local_cluster->httpAsyncClient().send(
```
Comment (joaopgrassi) [inline source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_fetcher.cc:34]: I guess they will probably mention the same case that we should either not care about the callback or keep the active requests somehow.

**Review by joaopgrassi — COMMENTED** [review]
Looks good and cool that it works already. I think the biggest concerns will be:

- How we keep track of the active requests (and cancel them when terminating Envoy, like raised in the HTTP exporter PR)
- What happens if the timer is ticked and there's an active request (server taking too long?)
- Should we consider having a default "return value"? It could be while the server is still responding with the sampler config, Envoy already received a request and needs to decide what to sample? 
- Should it stop trying if it's failing constantly?
- Should we make a first "sync" request during Envoy start up to see if it works and if not stop it from starting?

**Comment by samohte** [comment]
> Looks good and cool that it works already. I think the biggest concerns will be:
> 
>     * How we keep track of the active requests (and cancel them when terminating Envoy, like raised in the HTTP exporter PR)

At the moment the active request is stored in `active_request_`  which is cancelled in the destructor.
There can be only one active request, because the timer is re-started once the previous request is done.

>     * What happens if the timer is ticked and there's an active request (server taking too long?)

should not happen with the current implemenation.

>     * Should we consider having a default "return value"? It could be while the server is still responding with the sampler config, Envoy already received a request and needs to decide what to sample?

we will need a strategy what to do when we haven't received a config.

>     * Should it stop trying if it's failing constantly?

mh, not sure.

>     * Should we make a first "sync" request during Envoy start up to see if it works and if not stop it from starting?

I would not do that. Consider e.g. a network "hick-up" causes the first request to fail...

File: source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_fetcher.cc [hunk 2]
```diff
@@ -8,38 +8,35 @@ namespace Extensions {
...
 
-  for (const auto& header_value_option : http_service_.request_headers_to_add()) {
-    parsed_headers_to_add_.push_back({Http::LowerCaseString(header_value_option.header().key()),
-                                      header_value_option.header().value()});
-  }
+SamplerConfigFetcher::SamplerConfigFetcher(Server::Configuration::TracerFactoryContext& context,
+                                           const envoy::config::core::v3::HttpUri& http_uri,
+                                           const std::string& token)
+    : cluster_manager_(context.serverFactoryContext().clusterManager()), http_uri_(http_uri),
+      parsed_authorization_header_to_add_({Http::LowerCaseString("authorization"), token}),
```
Comment (joaopgrassi) [inline source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_fetcher.cc]: The token will be the "raw" one, so I think we should still add the header with value: `Api-Token {token}`.

File: test/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_fetcher_test.cc [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+// Test receiving a response with code != 200
+TEST_F(SamplerConfigFetcherTest, TestResponseErrorCode) {
+  SamplerConfigFetcher configFetcher(tracerFactoryContext_, http_uri_, "tokenXASSD");
+  timer_->invokeCallback();
+
+  Http::ResponseMessagePtr message(new Http::ResponseMessageImpl(
+      Http::ResponseHeaderMapPtr{new Http::TestResponseHeaderMapImpl{{":status", "401"}}}));
+  message->body().add("{\n \"rootSpansPerMinute\" : 4356 \n }");
```
Comment (joaopgrassi) [inline test/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_fetcher_test.cc:109]: I guess we can't test the request the config fetcher is making right? We could then check the path + token format to see if they are correct and will prevent tests from passing if someone changes those in the future

**Review by joaopgrassi — APPROVED** [review]
Makes sense to me! Just a few minor things

File: source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_fetcher.cc (hunk 2 above)
Comment (samohte) [reply source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_fetcher.cc]: done in 9a36c83acbb37a1fbad948e07356fc668a37a0c1

File: test/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_fetcher_test.cc (hunk 3 above)
Comment (samohte) [reply test/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_fetcher_test.cc:109]: I found a way to add a matcher which verifies the sent message, done in f636d1c839152080274ed0ca651d7beda8d86639

**Pull request merged**